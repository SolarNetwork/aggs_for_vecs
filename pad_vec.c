
Datum pad_vec(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(pad_vec);

/**
 * Pads a vector with NULLs out to n elements if necessary.
 * If the vector has more than n elements we raise an error.
 *
 * by Paul A. Jungwirth
 */
Datum
pad_vec(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  ArrayType *currentArray;
  int currentLength;
  Datum *currentVals;
  bool *currentNulls;
  int i;
  int arraySize;
  ArrayBuildState *state;
  Datum result;

  if (PG_ARGISNULL(0) || PG_ARGISNULL(1)) PG_RETURN_NULL();

  currentArray = PG_GETARG_ARRAYTYPE_P(0);
  arraySize = PG_GETARG_INT32(1);

  elemTypeId = ARR_ELEMTYPE(currentArray);
  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(currentArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &currentVals, &currentNulls, &currentLength);

  if (currentLength == arraySize) {
    PG_RETURN_ARRAYTYPE_P(currentArray);

  } else if (currentLength < arraySize) {
    state = initArrayResultWithNulls(elemTypeId, CurrentMemoryContext, arraySize);
    for (i = 0; i < currentLength; i++) {
      if (!currentNulls[i]) {
        state->dnulls[i] = false;
        state->dvalues[i] = datumCopy(currentVals[i], elemTypeByValue, elemTypeWidth);
      }
    }
    result = makeArrayResult(state, CurrentMemoryContext);
    PG_RETURN_DATUM(result);

  } else {
    ereport(ERROR, (errmsg("pad_vec found an array with %d elements but we're trying to pad out to only %d", currentLength, arraySize)));
  }
}
