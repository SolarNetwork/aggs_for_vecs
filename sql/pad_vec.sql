-- smallint pad same length
SELECT pad_vec(smallints, 3) FROM measurements WHERE sensor_id IN (4);
-- smallint pad adds NULLs
SELECT pad_vec(smallints, 5) FROM measurements WHERE sensor_id IN (4);
-- smallint pad too long
SELECT pad_vec(smallints, 2) FROM measurements WHERE sensor_id IN (4);

-- int pad same length
SELECT pad_vec(ints, 3) FROM measurements WHERE sensor_id IN (4);
-- int pad adds NULLs
SELECT pad_vec(ints, 5) FROM measurements WHERE sensor_id IN (4);
-- int pad too long
SELECT pad_vec(ints, 2) FROM measurements WHERE sensor_id IN (4);

-- bigint pad same length
SELECT pad_vec(bigints, 3) FROM measurements WHERE sensor_id IN (4);
-- bigint pad adds NULLs
SELECT pad_vec(bigints, 5) FROM measurements WHERE sensor_id IN (4);
-- bigint pad too long
SELECT pad_vec(bigints, 2) FROM measurements WHERE sensor_id IN (4);

-- real pad same length
SELECT pad_vec(reals, 3) FROM measurements WHERE sensor_id IN (4);
-- real pad adds NULLs
SELECT pad_vec(reals, 5) FROM measurements WHERE sensor_id IN (4);
-- real pad too long
SELECT pad_vec(reals, 2) FROM measurements WHERE sensor_id IN (4);

-- float8 pad same length
SELECT pad_vec(floats, 3) FROM measurements WHERE sensor_id IN (4);
-- float8 pad adds NULLs
SELECT pad_vec(floats, 5) FROM measurements WHERE sensor_id IN (4);
-- float8 pad too long
SELECT pad_vec(floats, 2) FROM measurements WHERE sensor_id IN (4);

-- numeric pad same length
SELECT pad_vec(nums, 3) FROM measurements WHERE sensor_id IN (4);
-- numeric pad adds NULLs
SELECT pad_vec(nums, 5) FROM measurements WHERE sensor_id IN (4);
-- numeric pad too long
SELECT pad_vec(nums, 2) FROM measurements WHERE sensor_id IN (4);
