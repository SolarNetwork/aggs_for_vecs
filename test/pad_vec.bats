load test_helper

@test "int16 pad same length" {
  result="$(query "SELECT pad_vec(ints, 3) FROM measurements WHERE sensor_id IN (4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "int16 pad adds NULLs" {
  result="$(query "SELECT pad_vec(ints, 5) FROM measurements WHERE sensor_id IN (4)")";
  echo $result;
  [ "$result" = "{1,NULL,2,NULL,NULL}" ]
}

@test "int16 pad too long" {
  run query "SELECT pad_vec(ints, 2) FROM measurements WHERE sensor_id IN (4)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2" ]
}

# @test "int32 count" {
  # result="$(query "SELECT vec_without_outliers('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 count" {
  # result="$(query "SELECT vec_without_outliers('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 count" {
  # result="$(query "SELECT vec_without_outliers('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 pad same length" {
  result="$(query "SELECT pad_vec(floats, 3) FROM measurements WHERE sensor_id IN (4)")";
  echo $result;
  [ "$result" = "{1,NULL,2}" ]
}

@test "float8 pad adds NULLs" {
  result="$(query "SELECT pad_vec(floats, 5) FROM measurements WHERE sensor_id IN (4)")";
  echo $result;
  [ "$result" = "{1,NULL,2,NULL,NULL}" ]
}

@test "float8 pad too long" {
  run query "SELECT pad_vec(floats, 2) FROM measurements WHERE sensor_id IN (4)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2" ]
}
