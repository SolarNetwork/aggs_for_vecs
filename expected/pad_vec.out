-- smallint pad same length
SELECT pad_vec(smallints, 3) FROM measurements WHERE sensor_id IN (4);
  pad_vec   
------------
 {1,NULL,2}
(1 row)

-- smallint pad adds NULLs
SELECT pad_vec(smallints, 5) FROM measurements WHERE sensor_id IN (4);
       pad_vec        
----------------------
 {1,NULL,2,NULL,NULL}
(1 row)

-- smallint pad too long
SELECT pad_vec(smallints, 2) FROM measurements WHERE sensor_id IN (4);
ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2
-- int pad same length
SELECT pad_vec(ints, 3) FROM measurements WHERE sensor_id IN (4);
  pad_vec   
------------
 {1,NULL,2}
(1 row)

-- int pad adds NULLs
SELECT pad_vec(ints, 5) FROM measurements WHERE sensor_id IN (4);
       pad_vec        
----------------------
 {1,NULL,2,NULL,NULL}
(1 row)

-- int pad too long
SELECT pad_vec(ints, 2) FROM measurements WHERE sensor_id IN (4);
ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2
-- bigint pad same length
SELECT pad_vec(bigints, 3) FROM measurements WHERE sensor_id IN (4);
  pad_vec   
------------
 {1,NULL,2}
(1 row)

-- bigint pad adds NULLs
SELECT pad_vec(bigints, 5) FROM measurements WHERE sensor_id IN (4);
       pad_vec        
----------------------
 {1,NULL,2,NULL,NULL}
(1 row)

-- bigint pad too long
SELECT pad_vec(bigints, 2) FROM measurements WHERE sensor_id IN (4);
ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2
-- real pad same length
SELECT pad_vec(reals, 3) FROM measurements WHERE sensor_id IN (4);
  pad_vec   
------------
 {1,NULL,2}
(1 row)

-- real pad adds NULLs
SELECT pad_vec(reals, 5) FROM measurements WHERE sensor_id IN (4);
       pad_vec        
----------------------
 {1,NULL,2,NULL,NULL}
(1 row)

-- real pad too long
SELECT pad_vec(reals, 2) FROM measurements WHERE sensor_id IN (4);
ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2
-- float8 pad same length
SELECT pad_vec(floats, 3) FROM measurements WHERE sensor_id IN (4);
  pad_vec   
------------
 {1,NULL,2}
(1 row)

-- float8 pad adds NULLs
SELECT pad_vec(floats, 5) FROM measurements WHERE sensor_id IN (4);
       pad_vec        
----------------------
 {1,NULL,2,NULL,NULL}
(1 row)

-- float8 pad too long
SELECT pad_vec(floats, 2) FROM measurements WHERE sensor_id IN (4);
ERROR:  pad_vec found an array with 3 elements but we're trying to pad out to only 2
